components/workflow.rst

```diff
@@ -19,6 +19,46 @@ You can install the component in 2 different ways:
 * :doc:`Install it via Composer </components/using_components>` (``symfony/workflow`` on `Packagist`_);
 * Use the official Git repository (https://github.com/symfony/workflow).
 
-For more information, see the code in the Git Repository.
+.. include:: /components/require_autoload.rst.inc
+
+Creating a Workflow
+-------------------
+
+The workflow component gives you an object oriented way to define a process
+or a life cycle that your object goes through. Each step or stage in the
+process is called a *place*. You do also define *transitions* to that describes
+the action to get from one place to another.
+
+.. image:: /_images/components/workflow/states_transitions.png
+
+A set of places and transitions creates a **definition**. A workflow needs
+a ``Definition`` and a way to write the states to the objects, (i.e. an
+instance of a :class:`Symfony\\Component\\Workflow\\MarkingStore\\MarkingStoreInterface`.
+
+Consider the following example for a blog post. A post can have places:
+'draft', 'review', 'rejected', 'published'. You can define the workflow
+like this::
+
+    use Symfony\Component\Workflow\Definition;
+    use Symfony\Component\Workflow\Transition;
+    use Symfony\Component\Workflow\Workflow;
+    use Symfony\Component\Workflow\MarkingStore\ScalarMarkingStore;
+
+    $states = ['draft', 'review', 'rejected', 'published'];
+
+    // Define a transaction with a name, where to go from and where to go to
+    $transitions[] = new Transition('to_review', 'draft', 'review');
+    $transitions[] = new Transition('publish', 'review', 'published');
+    $transitions[] = new Transition('reject', 'review', 'rejected');
+
+    $definition = new Definition($states, $transitions);
+    $definition->setInitialPlace('draft');
+
+    $marking = new ScalarMarkingStore('currentState');
+    $workflow = new Workflow($definition, $marking);
+
+The ``Workflow`` can now help you to decide what actions that are allowed
+on a blog post depending on what *place* it is in. This will keep your domain
```

Comment:
> If you write an issue with a summary of your thoughts with pros and cons, I'll be happy to submit a PR for that issue.

I can't do it, because the primary purpose of this Workflow is not obvious for me:
> It would be nice to describe the purpose of the component more.

The component is designed as very generic thing, so in theory it probably may have some interesting applications. However, provided [examples](https://github.com/lyrixx/SFLive-Paris2016-Workflow/blob/master/src/AppBundle/Entity/Article.php) don't impress. Compare it with rich domain model:
```php
final class Article
{
    private $id;
    private $title;
    private $approvedByJournalist = false;
    private $approvedBySpellchecker = false;
    private $published = false;

    public static function postDraft(int $articleId, ArticleTitle $title): Article
    {
        // ...
        $draft->apply(new DraftArticleWasPosted($articleId, $title));

        return $draft;
    }

    public function approveByJournalist()
    {
        if ($this->approvedByJournalist) {
            return;
        }

        $this->apply(new ArticleWasApprovedByJournalist($this->id));
        $this->publishIfItIsReady();
    }

    public function approveBySpellchecker() 
    {
        if ($this->approvedBySpellchecker) {
            return;
        }

        $this->apply(new ArticleWasApprovedBySpellchecker($this->id));
        $this->publishIfItIsReady();
    }

    public function changeTitle(ArticleTitle $newTitle)
    {
        if ($this->title->equals($newTitle)) {
            return;
        }

        $this->apply(new ArticleTitleWasChanged($this->id, $newTitle));
    }

    private function publishIfItIsReady()
    {
        if ($this->approvedByJournalist && $this->approvedBySpellchecker) {
            $this->apply(new ArticleWasPublished($this->id));
        }
    }

    // ...

    private function onArticleTitleWasChanged(ArticleTitleWasChanged $event)
    {
        // We have to approve new title
        $this->approvedByJournalist = false;
        $this->approvedBySpellchecker = false;
        $this->published = false;
    }
}
```

We see use-cases (`postDraft`, `approveBySpellchecker`, `approveBySpellchecker`, `changeTitle`), events (`ArticleWasApprovedByJournalist`, ..., `ArticleTitleWasChanged`) and it's not hard to understand lifecycle of the `Article` from this code. We lose this expressiveness.

Further, that sample project [checks permissions](https://github.com/lyrixx/SFLive-Paris2016-Workflow/blob/master/src/AppBundle/Workflow/GuardListener.php#L26) with `GuardEvent` hook:
```php
public function onTransitionJournalist(GuardEvent $event)
    {
        if (!$this->checker->isGranted('ROLE_JOURNALIST')) {
            $event->setBlocked(true);
        }
    }
```
... and how you can solve it with commands:

```php
final class ApproveArticleByJournalistHandler
{
    /**
     * @acl_role ROLE_JOURNALIST
     */
    public function handle(ApproveArticleByJournalist $command)
    {
        $blogPost = $this->blogPostRepository->find($command->getArticleId());
        $blogPost->approveByJournalist();
        $this->blogPostRepository->save($blogPost);
    }
}
```
And again we lose expressiveness by generic `GuardEvent` hook. It's harder to find it in the project, your have to search for string "workflow.article.guard.journalist_approval" instead of typing class name "`ApproveArticleByJournalis...`".

I would like to know if I'm wrong.